// @ideal-postcodes/core-interface/dist/client.js
"use strict";
/**
 * @module Client
 *
 * @description HTTP API Client
 */
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Client = exports.defaults = void 0;
/**
 * Default configuration
 */
exports.defaults = {
    tls: true,
    api_key: "",
    baseUrl: "api.ideal-postcodes.co.uk",
    version: "v1",
    strictAuthorisation: false,
    timeout: 10000,
    header: {
        Accept: "application/json",
        "Content-Type": "application/json",
    },
    tags: [],
    agent: {},
};
/**
 * Client Class
 */
var Client = /** @class */ (function () {
    function Client(config) {
        this.config = __assign(__assign({}, exports.defaults), config);
        this.config.header = __assign(__assign({}, exports.defaults.header), (config.header && config.header));
    }
    /**
     * Return base URL for API requests
     */
    Client.prototype.url = function () {
        var _a = this.config, baseUrl = _a.baseUrl, version = _a.version;
        return this.protocol() + "://" + baseUrl + "/" + version;
    };
    Client.prototype.protocol = function () {
        return this.config.tls ? "https" : "http";
    };
    return Client;
}());
exports.Client = Client;
