// @ideal-postcodes/core-interface/dist/index.js
"use strict";
/**
 * @module Exports
 *
 * @description Direct exports of core-interface
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.errors = exports.umprn = exports.udprn = exports.postcodes = exports.keys = exports.autocomplete = exports.addresses = exports.defaults = exports.Client = void 0;
var client_1 = require("./client");
Object.defineProperty(exports, "Client", { enumerable: true, get: function () { return client_1.Client; } });
Object.defineProperty(exports, "defaults", { enumerable: true, get: function () { return client_1.defaults; } });
__exportStar(require("./helpers"), exports);
__exportStar(require("./types"), exports);
var addresses = __importStar(require("./resources/addresses"));
exports.addresses = addresses;
var autocomplete = __importStar(require("./resources/autocomplete"));
exports.autocomplete = autocomplete;
var keys = __importStar(require("./resources/keys"));
exports.keys = keys;
var postcodes = __importStar(require("./resources/postcodes"));
exports.postcodes = postcodes;
var udprn = __importStar(require("./resources/udprn"));
exports.udprn = udprn;
var umprn = __importStar(require("./resources/umprn"));
exports.umprn = umprn;
var errors = __importStar(require("./error"));
exports.errors = errors;
