// @ideal-postcodes/core-interface/dist/resources/resource.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.listMethod = exports.retrieveMethod = void 0;
var util_1 = require("../util");
var error_1 = require("../error");
// Writes a resource to URL string
var toRetrieveUrl = function (options, id) {
    return [
        options.client.url(),
        options.resource,
        encodeURIComponent(id),
        options.action,
    ]
        .filter(function (e) { return e !== undefined; })
        .join("/");
};
var retrieveMethod = function (options) {
    var client = options.client;
    return function (id, request) {
        return client.config.agent
            .http({
            method: "GET",
            url: toRetrieveUrl(options, id),
            query: (0, util_1.toStringMap)(request.query),
            header: (0, util_1.toHeader)(request, client),
            timeout: (0, util_1.toTimeout)(request, client),
        })
            .then(function (response) {
            var error = (0, error_1.parse)(response);
            if (error)
                throw error;
            return response;
        });
    };
};
exports.retrieveMethod = retrieveMethod;
var listMethod = function (options) {
    var client = options.client, resource = options.resource;
    return function (request) {
        return client.config.agent
            .http({
            method: "GET",
            url: client.url() + "/" + resource,
            query: (0, util_1.toStringMap)(request.query),
            header: (0, util_1.toHeader)(request, client),
            timeout: (0, util_1.toTimeout)(request, client),
        })
            .then(function (response) {
            var error = (0, error_1.parse)(response);
            if (error)
                throw error;
            return response;
        });
    };
};
exports.listMethod = listMethod;
