// @ideal-postcodes/postcode-lookup/rollup.config.js
import commonjs from "@rollup/plugin-commonjs";
import resolve from "@rollup/plugin-node-resolve";
import ts from "@wessberg/rollup-plugin-ts";
import { dependencies } from "./package.json";

export const targets = ["ie 11", "last 2 versions"];

export const include = [
  ...Object.keys(dependencies).map((dep) => `node_modules/${dep}/**/*`),
];

export const polyfills = {
  Promise: "promise-polyfill",
  "Object.assign": "core-js-pure/features/object/assign",
};

export default [
  {
    input: "./lib/index.ts",
    output: {
      file: "./postcode-lookup.js",
      format: "umd",
      name: "IdealPostcodes",
      exports: "named", // Disable warning for default imports
    },
    plugins: [
      resolve({
        preferBuiltins: true,
        dedupe: ["@ideal-postcodes/jsutil"],
        mainFields: ["browser", "module", "main"],
      }),
      commonjs(),
      ts({
        transpiler: "babel",
        browserslist: targets,
        include: ["lib/**/*", "node_modules/core-js-pure/**/*", ...include],
        babelConfig: {
          presets: [["@babel/preset-env", { targets }]],
        },
      }),
    ],
  },
];
